addons/purchase/models/purchase.py

```diff
@@ -124,6 +124,13 @@ def _compute_invoice(self):
     company_id = fields.Many2one('res.company', 'Company', required=True, index=True, states=READONLY_STATES, default=lambda self: self.env.company.id)
     currency_rate = fields.Float("Currency Rate", compute='_compute_currency_rate', compute_sudo=True, store=True, readonly=True, help='Ratio between the purchase order currency and the company currency')
 
+    @api.constrains('company_id', 'order_line')
+    def _check_order_line_company_id(self):
+        for order in self:
+            companies = order.mapped('order_line.product_id.company_id')
+            if companies and companies != order.company_id:
+                raise ValidationError((_("Your quotation contains products from company %s whereas your quotation belongs to company %s. \n Please change the company of your quotation or remove the products from other companies.") % (companies.mapped('display_name'), order.company_id.display_name)))
+
```

Comment:
need to list the bad products in the error message to help the user correct the issue
``` python
bad_products = order.order_line.product_id.filtered(lambda p: p.company_id and p.company_id != order.company_id  # or something of the kind, untested
raise ValidationError(_("Your quotation contains products from company %s whereas your quotation belongs to company %s. \n Please change the company of your quotation or remove the products from other companies (%s).") % (companies.mapped('display_name'), order.company_id.display_name, ', '.join(bad_products.mapped('display_name'))))
```